schema: replace schema changer reflection with type switch

ecwall opened the pull request:
Fixes #110917

Currently the schema changer uses reflection to dispatch `tree.Statement` impls to the function that handles that impl.
https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L177

This circumvents the compiler and there are hand written checks that attempt to do validations that the compiler normally would.
https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L80

This PR replaces these reflection usages with type switch in order to leverage the compiler and make the code easier to read.

ecwall committed:
schema: replace supportedStatement.fn reflection with type switch

Informs #110917

Previously, `fn`'s type was interface{}, but had assertions in an init function
to make sure that it was actually a func that has inputs and output matching
certain types.

This commit uses a type switch and the compiler to get the same guarantees.

Release note: None

ecwall committed:
schema: replace supportedStatement.checks reflection with type switch

Informs #110917

Previously, `checks`'s type was interface{}, but had assertions in an
init function to make sure that it was actually a func that has inputs and
output matching certain types.

This commit uses a type switch and the compiler to get the same guarantees.

Release note: None

ecwall committed:
schema: replace supportedStatement.on reflection with type switch

Informs #110917

`on` is used to bypass the requirement that mode is UseNewSchemaChangerUnsafeAlways
or UseNewSchemaChangerUnsafe. It has been replaced by checking mode for "off" cases.

Release note: None

ecwall committed:
schema: initialize supportedStatementTags with supportedStatement.statementTag values

Previously, `statementTag` was only used to populate supportedStatementTags in an init
function.

Now supportedStatementTags is initialized as a map literal with the same statementTags.

Release note: None

ecwall committed:
schema: move supportedStatementTags into statement_control.go

`supportedStatementTags` is only used inside statement_control.go so it is now in
that file.

process_test.go has been renamed statement_control_test.go because it only contains
a test for statement_control.go.

Release note: None

ecwall committed:
schema: remove 22.1, 22.2 version checks

This commit removes the version checks for 22.1 and 22.2 because the 23.2
binary can never be in a mixed state with those versions.

Release note: None

ecwall committed:
schema: remove redundant mode check in alterTableChecks

alterTableChecks is only called by IsFullySupportedWithFalsePositive which
already checks that the mode is not UseNewSchemaChangerOff.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110922)
<!-- Reviewable:end -->

chrisseto commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go:
```diff
@@ -122,45 +35,55 @@ func IsFullySupportedWithFalsePositive(
+		return alterTableChecks(typedN, activeVersion)
+	case *tree.CreateIndex:
+		return activeVersion.IsActive(clusterversion.V23_1)
+	case *tree.DropDatabase:
```
nit: I've found that it's much more readable to combine cases that are the same into a single case. Once the line starts to get very long, go fmt will wrap it automatically.

```go
switch {
case *tree.DropOwnedBy, *tree.DropSchema, *tree.DropSequence, *tree.DropType:
     return true
}


```

chrisseto reviewed (commented):
I'm not sure I agree that this change, overall, is for the better.

For each individual type switch, I do agree that that statement is much more readable. 

However, there are now 3 different type switches and a map across 4 files that developers will need to update without any validation or linting to remind them. This implementation is much more likely to result in accidental mistakes. Even if go had exhaustive type switching, the usage of `default` makes it difficult to statically verify.

ecwall committed:
schema: merge filter and execution type switches

This commit combines the filter and execution type switches to prevent
accidentally adding a case to one, but not the other.

Release note: None

Xiang-Gu commented:
I'm inclined to agree with Chris' sentiment. The proposed changes seem to result in slight readability improvement but more maintenance overhead (and more prone to accidental mistakes). 
IIUC, the motivation is you want to get rid of [those 27 lines ](https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L79C1-L107) where we check callbacks are functions with expected signature. Is it really that bad?

ecwall commented:
> I'm not sure I agree that this change, overall, is for the better.
> 
> For each individual type switch, I do agree that that statement is much more readable.
> 
> However, there are now 3 different type switches and a map across 4 files that developers will need to update without any validation or linting to remind them. This implementation is much more likely to result in accidental mistakes. Even if go had exhaustive type switching, the usage of `default` makes it difficult to statically verify.

I added another commit to the filtering and processing type switches into one. Process now returns a function that can be called to do the actual processing.

ecwall commented:
> I'm inclined to agree with Chris' sentiment. The proposed changes seem to result in slight readability improvement but more maintenance overhead (and more prone to accidental mistakes).

Since he made that comment, I added another commit that merges some of the type switches.

> IIUC, the motivation is you want to get rid of [those 27 lines ](https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L79C1-L107) where we check callbacks are functions with expected signature. Is it really that bad?

I want to replace reflection in all of these places with more idiomatic type switch to handle a type union:

1. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L76
2. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L126
3. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L151
4. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L177
5. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L51
6. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L101
7. https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L176

In addition this change removes this [type alias](https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L35) of which the instantiations do not supply a statement tag even though it is on the original type.

Go provides type switches to solve the problem of different interface implementations needing different handling. If you include a declaration (`typedX := x.(type)`), it will even treat the variable as the type in the matching `case` to avoid type conversion statements. Before using reflection, we should determine that there are no other language features that can solve our problem IMO.

chrisseto reviewed (commented):
> In addition this change removes this [type alias](https://github.com/cockroachdb/cockroach/blob/97e4549307d44158d5d593032887452f518df63c/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L35) of which the instantiations do not supply a statement tag even though it is on the original type.

> Go provides type switches to solve the problem of different interface implementations needing different handling. If you include a declaration (`typedX := x.(type)`), it will even treat the variable as the type in the matching `case` to avoid type conversion statements. Before using reflection, we should determine that there are no other language features that can solve our problem IMO.

In general, I agree that the existing system is a bit obtuse and difficult to understand. If there was a bug or edge case that would be fixed by this refactor it'd be much easier to justify a refactor. However, the [originally referenced issue](https://github.com/cockroachdb/cockroach/issues/110917) doesn't mention such nor does it include input from anyone else.

Could we potentially have a couple people brain storm solutions and collectively decide on a direction rather than feel bound to a single solution because it exists as code? 

I'm still concerned about the lack of exhaustive type switching. The existing structure also requires a human being to interpret go code to determine if something is or is not supported by the DSC, which are known to be quite error prone when compared to computers.

One potential idea is to have an exhaustive test case (though I don't know how we could prove to ourselves that it is in fact exhaustive) which outputs the original table in a datadriven test for easy consumption?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110922#-:-NgLDeUQAwXWb9i-c04J:b-xl27iy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @Xiang-Gu)



<!-- Sent from Reviewable.io -->
